### PR TITLE
feat(source-scan): docker checks

### PR DESCRIPTION
**Pull request by dj8yfo**
fast-forward extension of #164 

1. [Cargo.toml](https://github.com/dj8yfo/sample_no_workspace/blob/1fd39bcbe4d523ac26c3d07c1f49ecc0fc65d9e7/Cargo.toml) => `sample-crate-12.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build",
        "--no-default-features",
        "--features",
        "near-sdk/expensive-debug"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "",
      "source_code_snapshot": "git+https://github.com/dj8yfo/sample_no_workspace.git?rev=1fd39bcbe4d523ac26c3d07c1f49ecc0fc65d9e7"
    },
    "link": "https://github.com/dj8yfo/sample_no_workspace/tree/1fd39bcbe4d523ac26c3d07c1f49ecc0fc65d9e7",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "0.7.7"
  }    
  ```
  
</p>
</details>

2. [Cargo.toml](https://github.com/dj8yfo/sample_workspace/blob/86aa6ccf4610f072cd207db86869e6324779780f/self-updates/update/Cargo.toml) => `sample-subpath-in-repo-6.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build",
        "--no-release",
        "--no-doc"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "self-updates/update",
      "source_code_snapshot": "git+https://github.com/dj8yfo/sample_workspace.git?rev=86aa6ccf4610f072cd207db86869e6324779780f"
    },
    "link": "https://github.com/dj8yfo/sample_workspace/tree/86aa6ccf4610f072cd207db86869e6324779780f",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "1.7.7"
  }
  ```
  
</p>
</details>

3. [dir containing submodule](https://github.com/dj8yfo/sample_workspace_with_submodules/tree/4b466921b89df67d6e8d1f9956b55a677edb2497/self-updates) => `sample-crate-in-submodule-5.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build",
        "--no-release",
        "--no-doc"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "self-updates/update",
      "source_code_snapshot": "git+https://github.com/dj8yfo/sample_workspace_with_submodules.git?rev=4b466921b89df67d6e8d1f9956b55a677edb2497"
    },
    "link": "https://github.com/dj8yfo/sample_workspace_with_submodules/tree/4b466921b89df67d6e8d1f9956b55a677edb2497",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "1.0.0"
  }
  ```
  
</p>
</details>

4. [Cargo.toml](https://bitbucket.org/dj8yfomule/scratch_check_git/src/finalization/Cargo.toml) => `bitbucket-scratch-4.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build",
        "--no-release",
        "--no-doc"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "",
      "source_code_snapshot": "git+https://feepdake27@bitbucket.org/dj8yfomule/scratch_check_git.git?rev=af9e835380053205f7fcc9b232a7ebb1c2c20694"
    },
    "link": "https://bitbucket.org/dj8yfomule/scratch_check_git/src/finalization/",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "0.1.0"
  }
  ```
  
</p>
</details>

5. factory [Cargo.toml](https://github.com/dj8yfo/factory-rust/blob/62a34ac86e6323ee6351b20fded7fc668b1e0277/factory/Cargo.toml) => `repro-fct-19.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build",
        "--no-default-features",
        "--features",
        "near-sdk/expensive-debug"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "factory",
      "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust.git?rev=62a34ac86e6323ee6351b20fded7fc668b1e0277"
    },
    "link": "https://github.com/dj8yfo/factory-rust/tree/62a34ac86e6323ee6351b20fded7fc668b1e0277",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "0.1.13"
  }
  ```
  
</p>
</details>

6. `repro-fct-19.testnet` => product `donation-product.repro-fct-19.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "product-donation",
      "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust.git?rev=62a34ac86e6323ee6351b20fded7fc668b1e0277"
    },
    "link": "https://github.com/dj8yfo/factory-rust/tree/62a34ac86e6323ee6351b20fded7fc668b1e0277",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "0.2.9"
  }
  ```
  
</p>
</details>

7. product [Cargo.toml](https://github.com/dj8yfo/factory-rust/blob/62a34ac86e6323ee6351b20fded7fc668b1e0277/product-donation/Cargo.toml) => `repro-fct-product-19.testnet`
<details>
  <summary>contract_source_metadata</summary><p>

  ```json
  {
    "build_info": {
      "build_command": [
        "cargo",
        "near",
        "build"
      ],
      "build_environment": "dj8yfo/sourcescan:0.x.x-dev-cargo-near-finalization@sha256:f86a095cb3daed21d0b5f86a4b5d9da0c0e26835f85c42678e1d460c9caa2c12",
      "contract_path": "product-donation",
      "source_code_snapshot": "git+https://github.com/dj8yfo/factory-rust.git?rev=62a34ac86e6323ee6351b20fded7fc668b1e0277"
    },
    "link": "https://github.com/dj8yfo/factory-rust/tree/62a34ac86e6323ee6351b20fded7fc668b1e0277",
    "standards": [
      {
        "standard": "nep330",
        "version": "1.2.0"
      }
    ],
    "version": "0.2.9"
  }
  ```
  
</p>
</details>

### CODE REVIEW

**Comment by dj8yfo** [comment]
no_bin_linux


no_bin_mac

File: cargo-near/src/commands/build_command/mod.rs [hunk 1]
```diff
@@ -7,16 +7,18 @@ use crate::{
...
-pub const INSIDE_DOCKER_ENV_KEY: &str = "CARGO_NEAR_BUILD_ENVIRONMENT";
-pub const BUILD_CMD_ENV_KEY: &str = "CARGO_NEAR_BUILD_COMMAND";
-pub const CONTRACT_PATH_ENV_KEY: &str = "CARGO_NEAR_CONTRACT_PATH";
-pub const SOURCE_CODE_SNAPSHOT_ENV_KEY: &str = "CARGO_NEAR_SOURCE_CODE_SNAPSHOT";
+pub const NEP330_INSIDE_DOCKER_ENV_KEY: &str = "CARGO_NEAR_BUILD_ENVIRONMENT";
+pub const NEP330_BUILD_CMD_ENV_KEY: &str = "CARGO_NEAR_BUILD_COMMAND";
+pub const NEP330_CONTRACT_PATH_ENV_KEY: &str = "CARGO_NEAR_CONTRACT_PATH";
+pub const NEP330_SOURCE_CODE_SNAPSHOT_ENV_KEY: &str = "CARGO_NEAR_SOURCE_CODE_SNAPSHOT";
 // ====================== End section =======================================
 pub const REPO_LINK_HINT_ENV_KEY: &str = "CARGO_NEAR_REPO_LINK_HINT";
```
Comment (frol) [inline cargo-near/src/commands/build_command/mod.rs:19]: See https://github.com/near/near-sdk-rs/pull/1178/files#r1628196147

File: cargo-near/src/commands/build_command/docker/git_checks/pushed_to_remote.rs [hunk 2]
```diff
@@ -16,7 +17,8 @@ pub fn check(git_url: &url::Url, commit_id: git2::Oid) -> color_eyre::Result<()>
         match repo {
             Ok(repo) => {
                 println!(" {}", "Checking if HEAD is present...".green());
-                repo.find_commit(commit_id)?;
+                repo.find_commit(commit_id)
+                    .wrap_err("commit wasn't found in remote repo")?;
```
Comment (frol) [inline cargo-near/src/commands/build_command/docker/git_checks/pushed_to_remote.rs:21]:
```suggestion
                    .wrap_err("commit wasn't found in remote repo. Please, push the changes to the remote repository so reproducible builds become possible")?;
```

File: cargo-near/src/commands/build_command/docker/docker_checks/hello_world.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        color_eyre::eyre::eyre!(ERR_SANITY),
+    )?;
+
+    if !output.status.success() {
+        let stderr = std::str::from_utf8(&output.stderr)?;
+        println!();
+        println!("{}", stderr.yellow());
+        if permission_denied(&output.status, stderr)? {
+            println!("{}", "Permission denied!".cyan());
+            super::print_installation_links(true);
```
Comment (frol) [inline cargo-near/src/commands/build_command/docker/docker_checks/hello_world.rs:23]: nit: This `bool` argument drives me crazy. Let's just extract that printing into a separate function and call it separately here. It is the only place where `true` is passed.

File: cargo-near/src/commands/abi_command/abi.rs [hunk 4]
```diff
@@ -93,6 +94,7 @@ pub(crate) fn generate_abi(
             ("CARGO_PROFILE_DEV_OPT_LEVEL", "0"),
             ("CARGO_PROFILE_DEV_DEBUG", "0"),
             ("CARGO_PROFILE_DEV_LTO", "off"),
+            (BUILD_RS_ABI_STEP_HINT_ENV_KEY, "true"),
```
Comment (frol) [inline cargo-near/src/commands/abi_command/abi.rs:97]: Do I see it right that it is not used anywhere? Let's remove dead code

**Review by frol — CHANGES_REQUESTED** [review]

File: cargo-near/src/commands/abi_command/abi.rs (hunk 4 above)
Comment (frol) [reply cargo-near/src/commands/abi_command/abi.rs:97]: It is actually in use: https://github.com/dj8yfo/cargo-near/blob/ac64571dbe1246175e2f5ba64582dfd7b439a74f/cargo-near/src/build_extended/build_script.rs#L44C46-L47

File: cargo-near/src/build_extended/build_script.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
+//! TODO: replace `cargo:` -> `cargo::`, as the former is being deprecated since rust 1.77
+//! or handle both with `rustc_version`
```
[the remark this answers is not in the document]
Comment (dj8yfo) [reply cargo-near/src/build_extended/build_script.rs]: Additional context: 

[original resolution pull](https://github.com/rust-lang/cargo/pull/12201)

[issue - cargo:: syntax should be permitted despite MSRV with the right configuration](https://github.com/rust-lang/cargo/issues/13868)
[comment - misleading docs](https://github.com/rust-lang/cargo/issues/13868#issuecomment-2101340733)
[doc - link around the misleading section](https://doc.rust-lang.org/nightly/cargo/reference/build-scripts.html#outputs-of-the-build-script)
Screenshot_20240614_154821


[comment - fact/assertion that old syntax won't be removed](https://github.com/rust-lang/cargo/issues/13868#issuecomment-2112320077)
[cfg_aliases source code](https://github.com/katharostech/cfg_aliases/blob/master/src/lib.rs#L368-L370)

**Review by frol — APPROVED** [review]
@dj8yfo Great job!

**Pull request merged**